src/NHibernate/Linq/Visitors/QuerySourceLocator.cs

```diff
@@ -37,16 +37,30 @@ public override void VisitAdditionalFromClause(AdditionalFromClause fromClause,
 			base.VisitAdditionalFromClause(fromClause, queryModel, index);
 		}
 
-        public override void VisitMainFromClause(MainFromClause fromClause, QueryModel queryModel)
-        {
-            if (_type.IsAssignableFrom(fromClause.ItemType))
-            {
-                _querySource = fromClause;
-            }
-            else
-            {
-                base.VisitMainFromClause(fromClause, queryModel);
-            }
-        }
-    }
-}
\ No newline at end of file
+		public override void VisitNhJoinClause(NhJoinClause joinClause, QueryModel queryModel, int index)
+		{
+			if (_type.IsAssignableFrom(joinClause.ItemType))
+			{
+				if (_querySource == null)
```

Comment:
Looks as an additional change, for which I do not get the rational. Is it fixing an untested case?